### PR TITLE
A Few Bug Fixes

### LINKED ISSUE #43: Server crashes with Cause: Access Violation
Log attached
[nslog30-12-2021 18-11-31.txt](https://github.com/R2Northstar/NorthstarMods/files/7794795/nslog30-12-2021.18-11-31.txt)

### LINKED ISSUE #153: Crash when changing titan mid embark animation
https://streamable.com/n8h75u

### LINKED ISSUE #161: SCRIPT ERROR: [SERVER] Given object is not an entity (type = null) -> armBadge.ClearParent()
***May*** *be a duplicate of #153 due to the seemingly identical error message but I cannot reproduce it on my dedicated server, only on a listening server.*

Server crashed in the middle of a lowpop (10 players) Skirmish round.
(Not quite) full log:  [temp.log](https://github.com/R2Northstar/NorthstarMods/files/7936092/temp.log)
Last few lines:
```
[14:14:31] [info] [SERVER SCRIPT] Requested replacement Titan from eye pos <1036.77, 1150.82, 212.299> view angles <-2.7
5151, 166.975, 0> player origin <1036.56, 1150.87, 158.299> map mp_rise
[14:14:31] [info] [SERVER SCRIPT] Dropping replacement titan at <-461.908, 1532.22, 264.246> with angles <0, 18.5712, 0>
[14:14:31] [info] [SERVER SCRIPT] burnmeter_amped_weapons
[14:14:31] [info] [SERVER SCRIPT] 1 30
[14:14:31] [info] [SERVER SPEW_MESSAGE] Setting player ---snip--- [8]" predicted first person proxy model to 
"models/weapons/arms/pov_titan_medium_ajax.mdl" (class "titan_atlas_stickybomb")
[14:14:31] [info] [SERVER SCRIPT] SCRIPT ERROR: [SERVER] Given object is not an entity (type = null)
[14:14:31] [info] [SERVER SCRIPT]  -> armBadge.ClearParent()
[14:14:31] [info] [SERVER SCRIPT][K
CALLSTACK
*FUNCTION [UpdateTitanArmBadge()] titan/sh_titan.gnut line [286]
*FUNCTION [CPlayer()] class/CPlayer.nut line [327]
*FUNCTION [CPlayer()] class/CPlayer.nut line [306]
*FUNCTION [PilotBecomesTitan()] mp/_titan_transfer.nut line [582]
*FUNCTION [TitanEmbark_PlayerEmbarks()] titan/sh_titan_embark.gnut line [986]

[14:14:31] [error] Northstar has crashed! a minidump has been written and exception info is available below:
[14:14:31] [error] Cause: Access Violation
Attempted to read from: 0x00000000
[14:14:31] [error] At: engine.dll + 0x247d3a
[14:14:31] [error]     Northstar.dll + 0x128d9e60
[14:14:31] [error]     ntdll.dll + 0x17002567d
[14:14:31] [error]     ntdll.dll + 0x1700548c5
[14:14:31] [error]     ntdll.dll + 0x17005245e
[14:14:31] [error]     engine.dll + 0x19607d3a
[14:14:31] [error]     engine.dll + 0x197d7466
[14:14:31] [error]     engine.dll + 0x194e0c2f
[14:14:31] [error]     engine.dll + 0x194e0542
[14:14:31] [error]     engine.dll + 0x1952db55
[14:14:31] [error]     engine.dll + 0x195888c0
[14:14:31] [error]     Northstar.dll + 0x128c18bc
[14:14:31] [error]     engine.dll + 0x1958652f
[14:14:31] [error]     engine.dll + 0x19750e2d
[14:14:31] [error]     engine.dll + 0x195879d9
[14:14:31] [error]     engine.dll + 0x1958793c
[14:14:31] [error]     launcher.dll + 0x12af5afd
[14:14:31] [error]     launcher.dll + 0x12af5afd
[14:14:31] [error]     launcher.dll + 0x12aed386
[14:14:31] [error]     NorthstarLauncher.exe + 0x140003bbd
[14:14:31] [error]     NorthstarLauncher.exe + 0x140006d40
[14:14:31] [error]     kernel32.dll + 0x7b62c599
[14:14:31] [error]     ntdll.dll + 0x170059d53
[14:14:31] [error]     NorthstarLauncher.exe + 0x0
[14:14:31] [error]     NorthstarLauncher.exe + 0x140006db0
[14:14:31] [error]     NorthstarLauncher.exe + 0x67ff0000
[14:14:31] [error]     NorthstarLauncher.exe + 0x0
[14:14:31] [error] RAX: 0x0
[14:14:31] [error] RBX: 0x19ba56d0
[14:14:31] [error] RCX: 0x0
[14:14:31] [error] RDX: 0x0
[14:14:31] [error] RSI: 0x2c49bb78
[14:14:31] [error] RDI: 0x2c49bb78
[14:14:31] [error] RBP: 0x2c49bd88
[14:14:31] [error] RSP: 0x7cf4a0
[14:14:31] [error] R8: 0x2c49bb78
[14:14:31] [error] R9: 0x65
[14:14:31] [error] R10: 0x2c49bd94
[14:14:31] [error] R11: 0x199c78ec
[14:14:31] [error] R12: 0x0
[14:14:31] [error] R13: 0x0
[14:14:31] [error] R14: 0xffffffff
[14:14:31] [error] R15: 0x1
0024:fixme:dbghelp_dwarf:dwarf2_read_range no entry found
0024:fixme:dbghelp_dwarf:dwarf2_read_range no entry found
0024:fixme:dbghelp:MiniDumpWriteDump NIY MiniDumpScanMemory
Error: Failed to run server: server exited: exit status 100.
```

Some deets:
* Running on PG's Docker image
* NS_EXTRA_ARGUMENTS: `-maxplayersplaylist +mpgamemode tdm +ns_private_match_last_map mp_forwardbase_kodai +ns_private_match_last_mode tdm +net_compresspackets 1 +net_compresspackets_minsize 64 +sv_maxrate 127000 +net_chan_limit_mode 2 +net_chan_limit_msec_per_sec 500 +ns_should_return_to_lobby 0 +ns_private_match_countdown_length 1 +ns_private_match_only_host_can_change_settings 2 +setplaylistvaroverrides "max_players 10 run_epilogue 0 fp_embark_enabled 1"`

### LINKED ISSUE #163: SCRIPT ERROR: [SERVER] Entity is null (soul.soul.lastOwner = player)
I've seen this happen more than 5 times in the last hour on my dedicated turbo_ttdm server. I'm not sure what the context is since I haven't been in-game for any of these occurrences.

<details>

```
northstar1_1  | te[01:06:29] [info] [SERVER SCRIPT] AttackerShouldTriggerReplay(): Not doing a replay by default..
northstar1_1  | _v[01:06:29] [info] [SERVER SCRIPT] AttackerShouldTriggerReplay(): Not doing a replay by default..
northstar1_1  | al[01:06:29] [info] [SERVER SCRIPT] AttackerShouldTriggerReplay(): Not doing a replay by default..
northstar1_1  | u[01:06:29] [info] [SERVER SCRIPT] AttackerShouldTriggerReplay(): Not doing a replay by default..
northstar1_1  | es?[01:06:29] [info] [SERVER SCRIPT] AttackerShouldTriggerReplay(): Not doing a replay by default..
northstar1_1  | id[01:06:29] [info] [SERVER SCRIPT] AttackerShouldTriggerReplay(): Not doing a replay by default..
northstar1_1  | =8f1[01:06:29] [info] [SERVER SCRIPT] Post impact,anim is done.
northstar1_1  | cb5657[01:06:29] [info] [SERVER SCRIPT] pilot primary skin/camo 1 151.
northstar1_1  | b[01:06:29] [info] [SERVER SCRIPT] pilot secondary skin/camo 1 151.
northstar1_1  | 4[01:06:29] [info] [SERVER SCRIPT] pilot weapon3 skin/camo 1 150.
northstar1_1  | 1d37096d3[01:06:29] [info] [SERVER SCRIPT] SCRIPT ERROR: [SERVER] Entity is null.
northstar1_1  | [01:06:29] [info] [SERVER SCRIPT]  -> soul.soul.lastOwner = player.
northstar1_1  | 5[01:06:29] [info] [SERVER SCRIPT].[K.
northstar1_1  | CALLSTACK.
northstar1_1  | *FUNCTION [PilotBecomesTitan()] mp/_titan_transfer.nut line [524].
northstar1_1  | *FUNCTION [RespawnAsTitan()] mp/_base_gametype_mp.gnut line [481].
northstar1_1  |
northstar1_1  | 3[01:06:29] [info] [SERVER SCRIPT] LOCALS.
northstar1_1  | [soul] NULL.
northstar1_1  | [fullCopy] true.
northstar1_1  | [titan] ENTITY (npc_titan [109] (NPC "npc_titan_auto_atlas_stickybomb" at <8416.26 -6640.95 4055.51>)).
```

(sorry about the screwed up log formatting)

</details>

### LINKED ISSUE #175: Server crash when player has negative XP and kills someone
As described in the [Discord FAQ](https://discord.com/channels/686167940994695169/686167940994695173/936320228793651290), it's possible to set your gen/XP using console commands.

If you set your XP to a negative value, you can crash any server by killing a player.

The FAQ mentions the following:
> Make sure to set your xp number to less than 472 to avoid a crash

That probably references a similar problem, so I'm guessing it's not a huge surprise that negative XP also causes issues. I think it's good to have an issue for it regardless.

### Some stuff related to negative XP
```
[2022-01-31 21:42:24.699] [info] [SERVER SPEW_MESSAGE] Setting player "player xaleelax1234 [1]" predicted first person proxy model to "" (class "spectator")
[2022-01-31 21:42:30.148] [info] [SERVER SPEW_WARNING] Client 1 has negative XP of -1000
[2022-01-31 21:42:30.148] [info] [SERVER SPEW_WARNING] Client 1 has negative generation of -1000
[2022-01-31 21:42:30.149] [info] [SERVER SPEW_MESSAGE] SERVER: Lunge_SetTargetEntity(): Setting lunge target to (null)
[2022-01-31 21:42:30.149] [info] [SERVER SPEW_MESSAGE] Setting player "player reslario [2]" predicted first person proxy model to "" (class "spectator")
[2022-01-31 21:42:30.149] [info] [SERVER SPEW_MESSAGE] Player reslario (2) is using prediction
[2022-01-31 21:42:30.302] [info] [SERVER SPEW_MESSAGE] Player reslario fully connected at time 48.645
[2022-01-31 21:42:30.302] [info] [SERVER SCRIPT] Player client script initialization complete: entity (2: player reslario [2])
```

### Last output before crash
```
[2022-01-31 21:42:54.764] [info] [SERVER SCRIPT] SCRIPT ERROR: [SERVER] XP -1000 too low
[2022-01-31 21:42:54.766] [info] [SERVER SCRIPT]  -> int oldLevel = GetLevelForXP( oldXp )
[2022-01-31 21:42:54.766] [info] [SERVER SCRIPT] 
CALLSTACK
*FUNCTION [HandleXPGainForScoreEvent()] _xp.gnut line [65]
*FUNCTION [AddPlayerScore()] mp/_score.nut line [84]
*FUNCTION [ScoreEvent_PlayerKilled()] mp/_score.nut line [113]
*FUNCTION [PlayerOrNPCKilled()] mp/_base_gametype.gnut line [784]
*FUNCTION [CodeCallback_OnPlayerKilled()] mp/_base_gametype_mp.gnut line [235]

[2022-01-31 21:42:54.794] [error] Northstar has crashed! a minidump has been written and exception info is available below:
[2022-01-31 21:42:54.794] [error] Cause: Access Violation
Attempted to read from: 0x00000000
[2022-01-31 21:42:54.795] [error] At: engine.dll + 0x247d3a
[2022-01-31 21:42:54.795] [error]     Northstar.dll + 0x7ffaa67d9e60
[2022-01-31 21:42:54.795] [error]     ntdll.dll + 0x7ffaf0d9bfea
[2022-01-31 21:42:54.795] [error]     ntdll.dll + 0x7ffaf0d55812
[2022-01-31 21:42:54.796] [error]     ntdll.dll + 0x7ffaf0dc730e
[2022-01-31 21:42:54.796] [error]     engine.dll + 0x7ffa10697d3a
[2022-01-31 21:42:54.796] [error]     engine.dll + 0x7ffa10867466
[2022-01-31 21:42:54.797] [error]     engine.dll + 0x7ffa10570c2f
[2022-01-31 21:42:54.797] [error]     engine.dll + 0x7ffa10570542
[2022-01-31 21:42:54.797] [error]     engine.dll + 0x7ffa105bdb55
[2022-01-31 21:42:54.797] [error]     engine.dll + 0x7ffa106188c0
[2022-01-31 21:42:54.798] [error]     Northstar.dll + 0x7ffaa67c18bc
[2022-01-31 21:42:54.798] [error]     engine.dll + 0x7ffa1061652f
[2022-01-31 21:42:54.798] [error]     engine.dll + 0x7ffa107e0e2d
[2022-01-31 21:42:54.798] [error]     engine.dll + 0x7ffa106179d9
[2022-01-31 21:42:54.799] [error]     engine.dll + 0x7ffa1061793c
[2022-01-31 21:42:54.799] [error]     launcher.dll + 0x7ffac1f65afd
[2022-01-31 21:42:54.799] [error]     launcher.dll + 0x7ffac1f65afd
[2022-01-31 21:42:54.799] [error]     launcher.dll + 0x7ffac1f5d386
[2022-01-31 21:42:54.800] [error]     NorthstarLauncher.exe + 0x7ff767ab3bbd
[2022-01-31 21:42:54.800] [error]     NorthstarLauncher.exe + 0x7ff767ab6d40
[2022-01-31 21:42:54.800] [error]     KERNEL32.DLL + 0x7ffaeff554e0
[2022-01-31 21:42:54.800] [error]     ntdll.dll + 0x7ffaf0d2485b
[2022-01-31 21:42:54.801] [error] RAX: 0x0
[2022-01-31 21:42:54.801] [error] RBX: 0x7ffa10c356d0
[2022-01-31 21:42:54.802] [error] RCX: 0x0
[2022-01-31 21:42:54.802] [error] RDX: 0x0
[2022-01-31 21:42:54.802] [error] RSI: 0x7ffa2352bb78
[2022-01-31 21:42:54.802] [error] RDI: 0x7ffa2352bb78
[2022-01-31 21:42:54.803] [error] RBP: 0x7ffa2352bd88
[2022-01-31 21:42:54.803] [error] RSP: 0x7698daef80
[2022-01-31 21:42:54.804] [error] R8: 0x7ffa2352bb78
[2022-01-31 21:42:54.804] [error] R9: 0x65
[2022-01-31 21:42:54.805] [error] R10: 0x7ffa2352bd94
[2022-01-31 21:42:54.805] [error] R11: 0x7ffa10a578ec
[2022-01-31 21:42:54.805] [error] R12: 0x0
[2022-01-31 21:42:54.805] [error] R13: 0x0
[2022-01-31 21:42:54.806] [error] R14: 0xffffffff
[2022-01-31 21:42:54.806] [error] R15: 0x1
```

### PR DESCRIPTION
**Pull request by MiloAkerman**
Resolves #161 
Resolves #153
Resolves #175 
Resolves #163 
Resolves #43 (kinda)

### CODE REVIEW

File: Northstar.Custom/mod/scripts/vscripts/titan/sh_titan.gnut [hunk 1]
```diff
@@ -0,0 +1,1276 @@
...
+	if ( model == $"" )
+		return
+	expect asset( model )
+
+	entity soulOwner = soul.GetBossPlayer()
+	if ( !IsValid( soulOwner ) )
+		return
+
+	entity armBadge = soul.soul.armBadge
+    if(armBadge == null)
```
Comment (BobTheBob9) [inline Northstar.Custom/mod/scripts/vscripts/titan/sh_titan.gnut:283]: could you change this to an !IsValid check?

**Review by BobTheBob9 — CHANGES_REQUESTED** [review]

**Comment by MiloAkerman** [comment]
Done

**Comment by MiloAkerman** [comment]
@BobTheBob9 is the syntax right? As barichello pointed out I don't know if this is a native issue or collateral from some Northstar mod because I wasn't able to find any reports for this specific crash.

**Comment by MiloAkerman** [comment]
Now also resolves #175

**Comment by MiloAkerman** [comment]
Seeing as I have no idea what the intended behavior should be for negative XP i decided to just go back to 0. Could also make it positive with `*= -1` but that seems a little scuffed

**Comment by MiloAkerman** [comment]
Resolves #163

**Comment by MiloAkerman** [comment]
Wtf why did the empty lines on `_xp.gnut` get scuffed

**Comment by MiloAkerman** [comment]
Now resolves one of the three issues in #43

**Comment by abarichello** [comment]
I don't think the fixes implemented in `_anim.gnut` and `_titan_transfer.nut` tackle the root of these problems, sure adding validity checks might prevent the game from crashing but that will probably move the crash somewhere else. Since that's the best we have for now I'll merge it anyways. Thanks

**Comment by MiloAkerman** [comment]
The change in `_anim.gnut` is more so a fix of an issue that should theoretically be native. I didn't add or remove any functionality, just changed a `null` check to a hopefully more in-depth `IsValid()` check

**Pull request merged**